orttraining/orttraining/python/orttraining_pybind_state.cc

```diff
@@ -760,6 +772,42 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
         return ortmodule_graph_builder->GetGraphInfo();
       });
 
+  // Provide a convenient and well-documented way to make a gradient graph.
+  // It's possible to get the gradient graph through ORTModule by leveraging some "private" fields and not-so-well-documented APIs, so we provide this explicit and tested way to get the gradient graph.
+  py::class_<PyGradientGraphBuilder> gradient_graph_builder(m, "GradientGraphBuilder", R"pbdoc(A utility for making a gradient graph that can be used to help train a model.)pbdoc");
+  // Set up methods to match the C++ `GradientGraphBuilder` interface.
+  gradient_graph_builder.def(py::init([](
+                                          const std::string& model_path,
+                                          const std::unordered_set<std::string>& y_node_arg_names,
+                                          const std::unordered_set<std::string>& x_node_arg_names,
+                                          const std::string loss_node_arg_name) {
+        auto file_path = ToPathString(model_path);
+        std::shared_ptr<Model> model;
+        auto logger = logging::LoggingManager::DefaultLogger();
+        ORT_THROW_IF_ERROR(Model::Load(file_path, model, nullptr, logger));
+        GradientGraphConfiguration gradient_graph_config{};
+        gradient_graph_config.set_gradients_as_graph_outputs = true;
+        // Save some objects, otherwise they get lost.
+        auto gradient_graph_config_ptr = std::make_unique<GradientGraphConfiguration>(gradient_graph_config);        
+        auto logger_ptr = std::make_unique<logging::Logger>(logger);
+
+        auto builder = std::make_unique<GradientGraphBuilder>(
+            &model->MainGraph(),
+            y_node_arg_names,
+            x_node_arg_names,
+            loss_node_arg_name,
+            *gradient_graph_config_ptr,
+            *logger_ptr);
+
+        return std::make_unique<PyGradientGraphBuilder>(std::move(builder), std::move(model), std::move(logger_ptr), std::move(gradient_graph_config_ptr));
+      }))
+      .def("build", [](PyGradientGraphBuilder* gradient_graph_builder) {
+        ORT_THROW_IF_ERROR(gradient_graph_builder->builder->Build());
+      })
+      .def("save", [](PyGradientGraphBuilder* gradient_graph_builder, const std::string& path) {
```

Comment:
Same comment as above, it is possible to get the onnx graph instead of saving it?

Reply:
That would be cool. I built this to match the `GradientGraphBuilder` in C++ which has this `build(..)`, `save(...)` interface but I also thought that it's good because once `build` is called, the graph can be manipulated with methods that don't exist yet. Yeah it sucks to force developers to use a file if it's not really needed but if you're calling this method, then you probably do want the file, eventually.

Reply:
I think it's good to support exporting to a file for now. We can add support in the future for exporting the model to other formats, such as a string so that it can be loaded by `onnx.load_model_from_string`.